modules/spark-wallet.nix

```diff
@@ -20,16 +22,49 @@ in {
         "The path of the clightning data directory.";
       '';
     };
+    onion-service = mkOption {
+      type = types.bool;
+      default = false;
+      description = ''
+        "If enabled, configures spark-wallet to be reachable through an onion service.";
+      '';
+    };
   };
 
   config = mkIf cfg.enable {
+    services.tor.enable = cfg.onion-service;
+    services.tor.hiddenServices.spark-wallet = mkIf cfg.onion-service {
+      map = [{
+        port = 80; toPort = 9737;
+      }];
+      version = 3;
+    };
     systemd.services.spark-wallet = {
       description = "Run spark-wallet";
       wantedBy = [ "multi-user.target" ];
-      requires = [ "clightning.service" ];
-      after = [ "clightning.service" ];
+      requires = builtins.concatLists [ [ "clightning.service" ] tor-service ];
+      after = builtins.concatLists [ [ "clightning.service" ]  tor-service ];
+      preStart = ''
+        if ! test -e ${dataDir}; then
+          mkdir -m 0770 -p '${dataDir}'
+          chown -R 'clightning:clightning' '${dataDir}'
+        fi
+        CMD="${pkgs.spark-wallet}/bin/spark-wallet --ln-path ${cfg.ln-path} -Q -k -c /secrets/spark-wallet-login"
+        ${optionalString cfg.onion-service
+          ''
+          echo Getting onion hostname
+          until ls -l /var/lib/tor/onion/spark-wallet/hostname; do sleep 1; done
+          CMD="$CMD --public-url http://$(cat /var/lib/tor/onion/spark-wallet/hostname)"
+          ''
+        }
+        echo $CMD > ${dataDir}/startscript.sh
+        echo Running $(cat ${dataDir}/startscript.sh)
+        chown -R 'clightning:clightning' ${dataDir}
+        chmod u+x ${dataDir}/startscript.sh
+      '';
       serviceConfig = {
-        ExecStart = "${pkgs.spark-wallet}/bin/spark-wallet --ln-path ${cfg.ln-path} -k -c /secrets/spark-wallet-login";
+        PermissionsStartOnly = "true";
+        ExecStart = "${pkgs.bash}/bin/bash ${dataDir}/startscript.sh";
```

Comment:
I would use pkg.writeScript here instead of generating a script inside of preStart. Can't the work of preStart not be done inside of ExecStart as well?

Reply:
Yes, that would be nicer. But we still need to do some work as root because the user doesn't have permissions to read `/var/lib/tor/onion/spark-wallet/hostname`.

Reply:
How about just copying & chowning hostname in preStart to ${dataDir}? Seems like an equal amount of work :)

But overall ACK, doesn't seem to hard to refactor later.

Reply:
> Yes, that would be nicer. But we still need to do some work as root because the user doesn't have permissions to read `/var/lib/tor/onion/spark-wallet/hostname`.

I am working on this fixing `nodeinfo` so it can be run as operator. If the hostnames are tor group readable, clightning user could be added to host.